### ISSUE TITLE
Start testing the parsing code

### ISSUE THREAD

**Issue by LinguList** [lingpy/pysign#6]
The code for parsing is now extremely large, but I understand of course, why, because this is just complex. So we should start testing it. this means, we see to which degree each if-else-construct is in fact executed, to avoid empty code that is never used.

For this purpose, I would first need a list of signs, nothing more. But ideally, you could, @justopower, add some basic things you want to "get" from them when parsing. E.g., you want to have a certain shape, so you could have a table:

sign | property | expected
--- | --- | ---
XXX | shape | thumbs-up

this table (csv or tsv) should be placed in the folder `tests/`.

**Comment by justopower** [comment]
@LinguList : Is this the kind of table you're looking for? Just to make sure before starting. This would be everything for one sign.


sign | property | expected
-- | -- | --
'   ' | symmetry | none
'   ' | initial position | none
'   ' | dominant['shape'] | ''
'   ' | dominant['orientation'] | ''
'   ' | dominant['contact'] | none
'   ' | dominant['location'] | ''
'   ' | dominant['movement'] | ''
'   ' | dominant['repitition'] | none
'   ' | nondominant['shape'] | none
'   ' | nondominant['orientation'] | none
'   ' | nondominant['location'] | none
'   ' | nondominant['contact'] | none
'   ' | nondominant['movement'] | none
'   ' | meta['handshape'] | none
'   ' | meta['orientation'] | none
'   ' | meta['contact'] | none
'   ' | meta['location'] | none
'   ' | meta['movement'] | none
'   ' | meta['rest'] | '   '

**Comment by LinguList** [comment]
Yes!

**Comment by justopower** [comment]
Ok, @LinguList  have a look at test-signs.tsv:
[https://github.com/lingpy/pysign/blob/master/tests/test-signs.tsv](url)

**Comment by LinguList** [comment]
| input                             | par1        | par2        | out                                              | expected                                   |
|:----------------------------------|:------------|:------------|:-------------------------------------------------|:-------------------------------------------|
|                              | dominant    | shape       | ['\ue002', '']                                   |                                           |
|                              | dominant    | orientation | ['\ue020\ue038', '']                             |                                          |
|                              | dominant    | contact     | ['', '']                                         |                                            |
|                              | dominant    | location    | ['\ue052', '']                                   |                                           |
|                              | dominant    | movement    | ['\ue089', '']                                   |                                           |
|                              | dominant    | repetition  | ['']                                             |                                            |
|                              | nondominant | shape       | ['', '']                                         |                                            |
|                              | nondominant | orientation | ['', '']                                         |                                            |
|                              | nondominant | location    | ['', '']                                         |                                            |
|                              | nondominant | contact     | []                                               |                                            |
|                              | nondominant | movement    | ['', '']                                         |                                            |
|                              | meta        | handshape   | []                                               |                                            |
|                              | meta        | orientation | []                                               |                                            |
|                              | meta        | contact     | []                                               |                                            |
|                              | meta        | location    | []                                               |                                            |
|                              | meta        | movement    | []                                               |                                            |
|                              | meta        | rest        |                                                  |                                            |
|           | dominant    | shape       | ['\ue002', '']                                   |                                           |
|           | dominant    | orientation | ['\ue020\ue03e', '']                             |                                          |
|           | dominant    | contact     | ['', '']                                         |                                            |
|           | dominant    | location    | ['\ue051\ue059', '']                             |                                          |
|           | dominant    | movement    | ['\ue090', '']                                   |                                           |
|           | dominant    | repetition  | ['']                                             |                                            |
|           | nondominant | shape       | ['\ue001', '']                                   |                                           |
|           | nondominant | orientation | ['\ue029\ue03c', '']                             |                                          |
|           | nondominant | location    | ['\ue059\ue052', '']                             |                                          |
|           | nondominant | contact     | []                                               |                                            |
|           | nondominant | movement    | ['\ue0af', '']                                   |                                           |
|           | meta        | contact     | []                                               |                                            |
|           | meta        | rest        |                                                  |                                            |
|     | dominant    | shape       | ['\ue002', '\ue002\ue010\ue00d']                 |                                           |
|     | dominant    | orientation | ['\ue020\ue03e', '']                             |                                          |
|     | dominant    | contact     | ['', '']                                         |                                            |
|     | dominant    | location    | ['\ue051\ue059', '']                             |                                          |
|     | dominant    | movement    | [['\ue090', '\ue0aa', 'simultaneous'], '']       | ['\ue090', '\ue0aa', 'simultaneous']       |
|     | dominant    | repetition  | ['']                                             |                                            |
|     | nondominant | shape       | ['\ue001', '']                                   |                                           |
|     | nondominant | orientation | ['\ue029\ue03c', '']                             |                                          |
|     | nondominant | location    | ['\ue059\ue052', '']                             |                                          |
|     | nondominant | contact     | []                                               |                                            |
|     | nondominant | movement    | ['\ue0af', '']                                   |                                           |
|     | meta        | contact     | []                                               |                                            |
|     | meta        | rest        |                                                  |                                            |
|        | dominant    | orientation | ['\ue020\ue038', '']                             |                                          |
|        | dominant    | movement    | ['\ue0aa', '']                                   |                                           |
|        | dominant    | repetition  | ['']                                             |                                            |
|        | nondominant | shape       | ['', '']                                         |                                            |
|        | nondominant | orientation | ['', '']                                         |                                            |
|        | nondominant | location    | ['', '']                                         |                                            |
|        | nondominant | contact     | []                                               |                                            |
|        | nondominant | movement    | ['', '']                                         |                                            |
|        | meta        | handshape   | []                                               |                                            |
|        | meta        | orientation | []                                               |                                            |
|        | meta        | location    | []                                               |                                            |
|        | meta        | movement    | []                                               |                                            |
|                 | dominant    | shape       | ['\ue001', '']                                   |                                           |
|                 | dominant    | orientation | ['\ue029\ue03e', '']                             |                                          |
|                 | dominant    | movement    | ['\ue096', '']                                   |                                           |
|                 | dominant    | repetition  | ['\ue0d9']                                       |                                           |
|                 | nondominant | shape       | ['', '']                                         |                                            |
|                 | nondominant | orientation | ['', '']                                         |                                            |
|                 | nondominant | location    | ['', '']                                         |                                            |
|                 | nondominant | contact     | []                                               |                                            |
|                 | nondominant | movement    | ['', '']                                         |                                            |
|                 | meta        | handshape   | []                                               |                                            |
|                 | meta        | orientation | []                                               |                                            |
|                 | meta        | location    | []                                               |                                            |
|                 | meta        | movement    | []                                               |                                            |
|                 | meta        | rest        |                                                  |                                            |
|                    | dominant    | contact     | ['', '']                                         |                                            |
|                    | dominant    | location    | ['\ue051\ue059', '']                             |                                          |
|                    | dominant    | movement    | [['\ue086\ue0ba', '\ue0aa', 'simultaneous'], ''] | ['\ue086\ue0ba', '\ue0aa', 'simultaneous'] |
|                    | dominant    | repetition  | ['']                                             |                                            |
|                    | nondominant | shape       | ['', '']                                         |                                            |
|                    | nondominant | orientation | ['', '']                                         |                                            |
|                    | nondominant | location    | ['', '']                                         |                                            |
|                    | nondominant | contact     | []                                               |                                            |
|                    | nondominant | movement    | ['', '']                                         |                                            |
|                    | meta        | handshape   | []                                               |                                            |
|                    | meta        | orientation | []                                               |                                            |
|                    | meta        | contact     | []                                               |                                            |
|                    | meta        | location    | []                                               |                                            |
|                    | meta        | movement    | []                                               |                                            |
|                    | meta        | rest        |                                                  |                                            |
|       | dominant    | shape       | ['\ue005\ue011\ue00c\ue070', '']                 |                                        |
|       | dominant    | orientation | ['\ue029\ue03e', '']                             |                                          |
|       | dominant    | location    | ['\ue075', ['\ue052', '\ue052\ue059']]           | ['\ue052', '\ue052\ue059']                 |
|       | dominant    | movement    | ['\ue08c', ['\ue08c']]                           | ['\ue08c', '\ue08c']                       |
|       | dominant    | repetition  | ['']                                             |                                            |
|       | nondominant | shape       | ['', '']                                         |                                            |
|       | nondominant | orientation | ['', '']                                         |                                            |
|       | nondominant | location    | ['\ue075', '']                                   |                                            |
|       | nondominant | contact     | []                                               |                                            |
|       | nondominant | movement    | ['', '']                                         |                                            |
|       | meta        | handshape   | []                                               |                                            |
|       | meta        | orientation | []                                               |                                            |
|       | meta        | movement    | []                                               |                                            |
|       | meta        | rest        |                                                  |                                            |
|         | dominant    | orientation | ['\ue027\ue038', '']                             |                                          |
|         | dominant    | location    | ['\ue04d', '']                                   | ['\ue04d', '\ue068']                       |
|         | dominant    | movement    | [['\ue086\ue0c6', '\ue0aa', 'simultaneous'], ''] | ['\ue086\ue0c6', '\ue0aa', 'simultaneous'] |
|         | dominant    | repetition  | ['']                                             |                                            |
|         | nondominant | shape       | ['', '']                                         |                                            |
|         | nondominant | orientation | ['', '']                                         |                                            |
|         | nondominant | location    | ['\ue068', '']                                   |                                            |
|         | nondominant | contact     | []                                               |                                            |
|         | nondominant | movement    | ['', '']                                         |                                            |
|         | meta        | handshape   | []                                               |                                            |
|         | meta        | orientation | []                                               |                                            |
|         | meta        | movement    | []                                               |                                            |
|         | meta        | rest        |                                                  |                                            |
|                  | dominant    | shape       | ['\ue004\ue011\ue00d', '']                       |                                         |
|                  | dominant    | orientation | ['\ue029\ue03d', '']                             |                                          |
|                  | dominant    | contact     | ['', '']                                         |                                            |
|                  | dominant    | location    | ['\ue066', '']                                   |                                           |
|                  | dominant    | movement    | [['\ue089', '\ue0a4', 'simultaneous'], '']       | ['\ue089', '\ue0a4', 'simultaneous']       |
|                  | dominant    | repetition  | ['']                                             |                                            |
|                  | nondominant | shape       | ['', '']                                         |                                            |
|                  | nondominant | orientation | ['', '']                                         |                                            |
|                  | nondominant | location    | ['\ue068', '']                                   |                                           |
|                  | nondominant | contact     | []                                               |                                            |
|                  | nondominant | movement    | ['', '']                                         |                                            |
|                  | meta        | handshape   | []                                               |                                            |
|                  | meta        | orientation | []                                               |                                            |
|                  | meta        | contact     | []                                               |                                            |
|                  | meta        | movement    | []                                               |                                            |
|                  | meta        | rest        |                                                  |                                            |
|                     | dominant    | orientation | ['\ue029\ue03e', '']                             |                                          |
|                     | dominant    | contact     | ['\ue0d0\ue068', '']                             |                                          |
|                     | dominant    | location    | ['\ue04d', '']                                   |                                           |
|                     | dominant    | movement    | [['\ue0aa', '\ue089', '\ue0d9'], '']             | ['\ue0aa', '\ue089', '\ue0d9']             |
|                     | dominant    | repetition  | ['']                                             |                                            |
|                     | nondominant | shape       | ['', '']                                         |                                            |
|                     | nondominant | orientation | ['', '']                                         |                                            |
|                     | nondominant | location    | ['', '']                                         |                                            |
|                     | nondominant | contact     | []                                               |                                            |
|                     | nondominant | movement    | ['', '']                                         |                                            |
|                     | meta        | handshape   | []                                               |                                            |
|                     | meta        | orientation | []                                               |                                          |
|                     | meta        | location    | []                                               |                                            |
|                     | meta        | movement    | []                                               |                                            |
|                     | meta        | rest        |                                                  |                                            |

**Comment by LinguList** [comment]
This helps us already. see `tests/test_parse.py`. You find a data-list, and you find the params we pass, and the output. You see the error report above (just run: `$ python test_parse.py`). We can with tthis also run a real test, this shows, what lines of the code are NOT used so far, which is important for making the algorithm safe. Here, the data so far cover some 33%, so much more to do. 

But can I ask you first, @justopower, to have a look and see to which degree this can be adjusted, so that the tests run? You have a better overview on the mismatches here (if they are systematic or not).

**Comment by justopower** [comment]
Ok, I see that at some points, the values in 'expected' column are not correct. That must have been my mistake when creating [examples/test-signs.tsv](https://github.com/lingpy/pysign/blob/master/examples/test-signs.tsv). I will fix that. 

But two other things @LinguList : 1) I don't see any error report when I run test_parse.py; 
2) It is not unexpected about the 33%, if I understand you correctly, because some transcriptions can become very complex, but I kept the tests relatively simple at first.

**Comment by LinguList** [comment]
To see the errors, you have to have pytest installed:

```
$ pip install pytest pytest-cov
```

And then run:

```
$ pytest test_parse.py
```

Then, to see coverage:

```
$ pytest test_parse.py --cov=pysign 
```

And a detailed output

```
$ pytest test_parse.py --cov=pysign --cov-report=html
```

Then you have an HTML file where you see all results (but do not add them to git ;)

**Comment by justopower** [comment]
Ok, there are several functions in parse.py, not just parse_hamnosys(). Part of the "missing" lines in the coverage report tells us that these other functions are not used. But we're not trying to test these other functions so that is unhelpful.

There are many "missing" lines in parse_hamnosys too, but I think this has a lot to do with the fact that the test cases in [examples/test-signs.tsv](examples/test-signs.tsv) are simple examples; and unfortunately a lot of the code covers all of the marginal cases that would otherwise cause the parser to make mistakes. So this percentage will stay low until we test more complex transcriptions.

Other than that, @LinguList , I am not sure how to interpret the assertion error: `AssertionError: assert '' == None`

There are many lines in the variable `data` in `parse.py` that have `None` as the third value. How do we know which of these is causing the error? Should we simplify [examples/test-signs.tsv](examples/test-signs.tsv), just using one transcription at a time to isolate which is causing the problem, or can you see this already? I haven't been able to locate that information.

**Comment by LinguList** [comment]
Of course, we need to add more tests, but first, I wanted to have explicit examples for test-signs.tsv, which I by the way, included in the file as a list, as you have seen, right?

If you have cases like ```assert '' == None```, this means you need to replace `''` by `None`. So you need to adjust the test data (now in the `data` list).

The procedure we now need is: 

1. make sure the tests work (for that, to find out which line is not working, use `python test_parse.py` (with detailed error output)
2. use pytest for checking the code, but not now, it is a bit too early for that
3. add more tests, to make sure the most complex examples are also covered, and we learn if the code is doing what it is supposed to do

**Comment by justopower** [comment]
> Of course, we need to add more tests, but first, I wanted to have explicit examples for test-signs.tsv, which I by the way, included in the file as a list, as you have seen, right?
> 
Ha, yes of course. I can just test each one that way. Sorry!

**Comment by justopower** [comment]
Ok, there are two problems. First, we have to replace the other way around: not `''` by `None`, but rather the test doesn't like all of the `None`s in `data`, so they have to be switched to `''`. That is easy.

Second possibly requires work on the function itself `parse_hamnosys`: if you just use the function to parse, say, the first transcription (   ), you will see that some of the values under "nondominant" and "meta" return empty lists, which causes the following error: IndexError: list index out of range. So I think it would require that we either change the test or change `parse_hamnosys`. What do you recommend, @LinguList ?

**Comment by LinguList** [comment]
If an empty list is our expected output, and not `None`, we should
adjust the test. The problem is that we have two kinds of output: one
with a list, and one without, but this was not clear from the test data
you provided. For this reason, I just used the most frequently appearing
output (a list). But we can adjust the data accordingly that it also
indicates what output to expect.

**Comment by justopower** [comment]
Yes, you're right, of course @LinguList . It is partly to do with me not having experience with these tests and not realizing how to represent exactly what the function returns in the "expected" column.

So anyway, this is not difficult to change, I think.
- I will just run the function on each transcription separately and then re-do `test_signs.tsv`.
- Then we can adjust the test function, in particular the line `assert out[b[0]][b[1]][0] == c` is causing the issues now. 
- Then the tests should all run anyway and we go from there.

**Comment by LinguList** [comment]
@justopower, I prefer not to use a tsv file as input, as this is difficult for representing code. So it is easier to just leave the data in the script.

In fact, we can make a simple if/else: we add a column to the data, say, column 0, and then check like:

```
for a, b, c, d in data:
    out = parse_hamnosys(b)
    if a == 1:
        if out[c[0]][c[1]] == d:
            ...
```
You see?

**Comment by LinguList** [comment]
Do you want to go ahead with this?

**Comment by justopower** [comment]
Yes, sure. I can work with the script until it works, if that's what you're asking.

**Comment by LinguList** [comment]
Yep, and you just let me know if you encounter problems.

**Comment by justopower** [comment]
Ok, should work now, @LinguList . Have a look when you get the chance. 

If everything is ok, I can add more tests to `data`.

**Comment by justopower** [comment]
@LinguList Ok, I've finally had a chance to get back to the parsing code. Basically, I've just made sure that it still runs fine on the transcriptions that we have in `data` under [tests/test_parse.py](https://github.com/lingpy/pysign/blob/master/tests/test_parse.py).

So the parsing code works fine on those transcriptions. Maybe this is ok for now, and we can look again in case anything breaks when doing comparisons of signs. What do you think?
